Enable config setting `sparse_interface` to control sparray and spmatrix creation

dschult opened the pull request:
This PR sets up a config parameter `sparse_interface` to indicate "sparray" or "spmatrix" outputs, as suggested in #26418.

The first commit sets everything up and implements the system for a few modules. Please take a look and provide feedback for whether this is the way to proceed. The next commit(s) will implement these same style of changes throughout the library. If you would prefer they be in separate PRs let me know. (I'll keep Draft status until there is feedback and the full library is convered.)

More specifically, this PR does the following:
- adds `sparse_interface` to the config parameters. (I think this name is better than sparse_format because "format" means csr/coo/lil, etc in the sparse world.) The values it can hold are "sparray" or "spmatrix". Update config tests accordingly.
- adds `utils._sparse.py` with (private) helper functions. The difference is how much checking is done. Tests added too.
    -  `_as_sparse(x_sparse)`,  raises unless sparse input. converts to interface chosen by config.
    - `_select_interface_if_sparse(x)`, allows dense input with no action, sparse input uses `_as_sparse`.
    -  one-line convenience functions: `_convert_from_spmatrix_to_sparray(x)` and `_convert_from_sparray_to_spmatrix(x)`
- updates the following modules to use this helper utility to return or store newly created sparse objects.
    - `sklearn/feature_selection/text.py` and adapts tests. 
    - `sklearn/linear_model/_coordinate_descent.py` no tests change needed.
    - `sklearn/manifold/_locally_linear.py` no tests change needed.

@thomasjpfan can you see if this does what you had in mind? I tried to pick modules that cover returning sparse, setting estimators to hold sparse, and transforming to sparse, so you can see how this would work.

Let me know if you think `_as_sparse` should be a public function, and if my approach aligns with how you want it. The next steps for this PR are to repeat these type of changes throughout the library.

dschult committed:
Use sparse sparray by default while allowing spmatrix inputs

set up function _align_api_if_sparse() with tests
Also functions _ensure_sparse_index_int32() and safely_cast_index_arrays()
and _sparse_eye, _sparse_diags, _sparse_random to span Scipy <1.12 changes
Introduce SCIPY_VERSION_BELOW_1_12 and SCIPY_VERSION_BELOW_1_15 boolean flags
fix rng keyword arg for old SciPy versions
ensure 2d sparse
convert benchmarks
doc modules
convert csr_matrix to csr_array and CSR, COO, DIA, etc.
make doctests pass
pass tests on older scipy versions. i.e. ensure int32 indices where needed.
make it work for SciPy 1.8
improve test coverage

dschult commented:
Another note:
The new sparse construction function `eye_array(n)` which is the sparray version of `eye(n)` was released in SciPy v1.12 along with other construction functions like `diags_array`. So they will not work with the oldest supported version being v1.8. 

We can work around it for now with e.g. `_as_sparse(eye(n))`, but it will need to be updated later (before spmatrix is removed). 

The recent features for sparse by version are:
- v1.12 added construction function e.g. eye_array, diags_array, etc
- v1.14 added 1D sparray support
- v1.15 added indexing for sparray which returns 1D objects (like numpy.array does), e.g. `A[3,:] -> 1D array`
goals: v1.16 nD support,   v1.17 broadcasting of binary operations.

This info might help us decide when to support which versions. I think the construction functions are all that is currently needed. If/when we start using indexing code for both sparse and dense, we will likely want 1.15. If/when we want nD sparse we will need v1.16, and broadcasting binary operations in v1.17. But for now, 1.8 leaves out only construction functions from current code.

dschult commented:
It sounds like the community has decided to convert to SciPy sparray using the config parameter. This PR is a start toward that. I've removed the "Draft status" from the PR. I am ready to implement this approach in other parts of the library.

I have two questions:
- should I put the changes for other parts of the library in a different PR?
- Is there a timeline for bumping up the minimum version dependence on scipy? I will likely want to add a few functions to `fixes.py` and knowing their expected lifetime would give context to that effort.  If it isn't known, that's a fine answer too.

thomasjpfan commented on sklearn/feature_extraction/tests/test_text.py:
```diff
@@ -1611,7 +1612,12 @@ def test_tfidf_transformer_copy(csr_container):
-    assert X_transform is X_csr
+    # allow for config["sparse_interface"] to change output type
+    # there should be no data copied, but the `id` will change.
+    if _as_sparse(X_csr) is X_csr:
+        assert X_transform is X_csr
+    else:
+        assert X_transform is not X_csr
```
Although the object identity is not the same, is the underlying data/indices/indexes pointing to the same data? If so, we can check the data?

dschult replied:
Yes -- the underlying data is the same. Good idea.  I've added a check for `indptr`.

thomasjpfan commented on sklearn/utils/_sparse.py:
```diff
@@ -0,0 +1,62 @@
+from .._config import get_config
+
+
+def _as_sparse(X_sparse):
```
Do you think it'll be simpler to have `_as_sparse` be `_select_interface_if_sparse`?

dschult replied:
I guess the difference between function is whether we know already that the input is sparse. We usually know whether it is sparse. But there are some cases where the input could be dense or sparse.

Looking at this again now, the cost of a simpler one-function approach is small. We always check `issparse` so we could forego the exception and just pass-through anything that is not sparse.

Do you have a suggestion for the name of the single function? `_as_sparse` suggests it converts everything to sparse. But `_select_interface_if_sparse` is a long name. :)  Maybe `_align_api_if_sparse`?

thomasjpfan replied:
I'm okay with `_align_api_if_sparse`.

thomasjpfan reviewed (commented):
Yup, this is basically what I had in mind.

dschult commented on sklearn/semi_supervised/_label_propagation.py:
```diff
@@ -463,7 +463,15 @@ class distributions will exceed 1 (normalization may be desired).
         normalizer = affinity_matrix.sum(axis=0)
         if sparse.issparse(affinity_matrix):
-            affinity_matrix.data /= np.diag(np.array(normalizer))
+            normalizer = np.array(normalizer)  # convert np.matrix to np.array
+            if normalizer.ndim == 2:
+                # old spmatrix treatment. RHS is a scalar (b/c normalizer is 2D row)
+                affinity_matrix.data /= np.diag(normalizer)
+            else:
+                # We could use the (questionable) spmatrix treatment using:
+                # affinity_matrix.data /= np.diag(np.array(normalizer[np.newaxis, :]))
+                # Instead: use  numpy treatment dividing each row by its normalizer.
+                affinity_matrix.data /= normalizer[affinity_matrix.indices]
```
This change deserves a comment -- and maybe even a separate PR.
`normalizer` is the `axis=0` sum of `affinity_matrix`. When affinity_matrix is numpy or sparray this is a 1D array.
But when affinity_matrix is spmatrix, it is 2D.  So sparse matrices currently (on main) get special treatment:
```affinity_matrix.data /= np.diag(np.array(normalizer))```
normalizer is an `np.matrix` (2D with 1 row).  `np.array(normalizer)` is an array (still 2D with 1 row).  So np.diag of that is just the first entry in the first row (a scalar). 

But that divides the entire affinity_matrix by the sum of the first row!  So, instead of normalizing each row by its sum, we get the whole matrix normalized by the sum of the first row.  I think this is an error.  

When `affinity_matrix` is a numpy array, we do divide each row by its sum. So, currently sparse and dense compute different normalizations here.

I have added sparse array handling here to divide each row by its sum. But I have left the spare matrix code there in case backward compatibility is an issue. 

Should I open a separate PR for changing the spmatrix behavior so it doesn't divide by a scalar? Or just fix it in this PR? Or am I fixing it incorrectly in some way?

adrinjalali replied:
probably easier to review as a part of a separate PR.

dschult commented:
Okay, I think this is finally ready for review. I've implemented a config to determine the type of sparse matrix to return (sparray or spmatrix). Both kinds can be used for sparse inputs.

New helper functions in `utils._sparse.py`:
- `_align_api_if_sparse(X)` changes the sparse type if needed
- `_sparse_random`, `_sparse_eye_array`, `_sparse_diags`  (sparse construction functions across versions `1.8->1.12-> 1.15+`)
- `_ensure_sparse_index_int32(X)` sets index dtype for sparse X to be int32 if safe to do so. Raise if not. Useful when using libraries that only support int32 index arrays.
- `safely_cast_index_arrays' local version of `scipy.sparse._sputils.safely_cast_index_arrays` that works across SciPy versions.

New Scipy Version bool flags to ease indexing code that depends on SciPy version (mostly index that make 1D arrays):
- `SCIPY_VERSION_BELOW_1_12`
- `SCIPY_VERSION_BELOW_1_15`

All internal constructors are switched to `sparray`, e.g. `csr_array`, `csc_array`, etc.
Return values are converted to the interface specified in config 'sparse_interface'.
code, docs, rst files, doc_strings, benchmarks are all updated.

If you want me to separate this into smaller PRs, let me know.

CC: @thomasjpfan , @lorentzenchr, @jjerphan, #26418

adrinjalali commented on sklearn/_config.py:
```diff
@@ -193,6 +195,16 @@ def set_config(
+        The sparse interface used for every sparse object that scikit-learn produces,
+        e.g., function returns, estimator attributes, estimator properties, etc.
+
+        - `"sparray"`: Return sparse as SciPy sparse array
+        - `"spmatrix"`: Return sparse as SciPy sparse matrix
```
are we changing this default at some point? Should we introducing a deprecation cycle at the same time?

adrinjalali commented on sklearn/_config.py:
```diff
@@ -193,6 +195,16 @@ def set_config(
+        - `"sparray"`: Return sparse as SciPy sparse array
+        - `"spmatrix"`: Return sparse as SciPy sparse matrix
+
+        .. versionadded:: 1.7
```
```suggestion
        .. versionadded:: 1.8
```

adrinjalali reviewed (commented):
Seems pretty straightforward to me.

thomasjpfan commented on sklearn/manifold/_isomap.py:
```diff
@@ -297,6 +298,7 @@ def _fit_transform(self, X):
                 **self.nbrs_.effective_metric_params_,
             )
 
+        _ensure_sparse_index_int32(nbg)
```
How did you run into this issue? (And know that you need to ensure int32)

dschult replied:
Many tests fail in `manifold/tests/test_isomap.py` with SciPy v1.12. For example `sklearn/manifold/tests/test_isomap.py::test_isomap_simple_grid`.  
They do pass under SciPy 1.8 and SciPy 1.16. I ran the test suite for SciPy v1.8, v1.12 and v1.16 on this branch. And the v1.12 environment caused these failures.

I know that recent SciPy versions (I think 1.14+) already ensure index arrays are `int32` because I added that. But I don't know the history for why that isn't needed for SciPy v1.8.

dschult committed:
update versionadded for sparse_interface config

dschult committed:
tweak label_propagation to give same answer for spmatrix and sparray

dschult committed:
add scipy version checking to normalizer syntax. New versions match sparse and dense syntax.

dschult committed:
fix of recent update indexing sparse in a way not passing SciPy 1.10 min_dependency

lesteve commented:
> One thing that is not clear to me: how do we plan to change the default behaviour from using sparse matrix to sparse array? It feels like we can not use our standard strategy "let's warn for two releases and then change it"?

This was discussed at a bi-weekly meeting, and the consensus for now is to do something similar to our usual strategy for estimator parameters. When the config is accessed with the key `sparse_interface` and its value has not been explicitly by the user, we warn that the default will change and set it explicitly if you care (either `spmatrix` if you want backward-compat or `sparray` to have the future behaviour).

I think introducing this warning can be done in a separate PR.

The timeline for the deprecation hasn't been discussed. My feeling is that this could have tricky side-effects, so maybe the usual 2 release warnings and then change is a bit too short for such a change? This doesn't have to be decided before this PR is merged either.

dschult committed:
Merge branch 'main' into impl_as_sparse_function

dschult committed:
fix linting errors that appeared after merge to main

dschult committed:
Add first draft of release note.

dschult commented:
I went ahead and created a draft release note. Hopefully that is helpful for whoever needs to write the final version.

lorentzenchr reviewed (commented):
Partial review

lorentzenchr reviewed (commented):
Another bit

dschult committed:
correct test leading to avoiding all SCIPY version cruft

dschult committed:
check for data/indices not copied by looking at their .base attribute

dschult committed:
update whats_new changelog entry

dschult committed:
update pyamg import checking

dschult committed:
moved import of utils private function to top of file.

dschult reviewed (commented):
Thanks for the helpful review!
I've addressed the review comments -- I still have one question:
- Is there a general rule to follow for which functions to import at the top?

I move the import of `sklearn.utils._sparse._ensure_sparse_index_int32` to the top of the file. But that makes me wonder whether I need to import all `sklearn` tools at the top of the file. Some of them are public api, so presumably not those, e.g. `sklearn.config_context`. But what about e.g. `sklearn.utils._align_api_if_sparse`.

dschult committed:
Merge branch 'main' into impl_as_sparse_function

dschult committed:
Merge branch 'main' into impl_as_sparse_function

dschult committed:
rename whats_new rst file

lorentzenchr reviewed (approved):
@dschult Nice work!

dschult committed:
Merge branch 'main' into impl_as_sparse_function

dschult committed:
Merge branch 'main' into impl_as_sparse_function

dschult committed:
add suggestion for doc_string in _sparse.py

dschult commented:
I have merged with main and added the doc_string suggestion. 
Should be ready for review --  Thanks @lorentzenchr !

dschult committed:
Merge branch 'main' into impl_as_sparse_function

dschult commented:
Happy New Year!
Congrats on 1.8 release!

Let's get these (#31177) sparray vs spmatrix changes in before other major changes come along! :)

We've got one approval and need one more. Many of you have looked through this and given good feedback! (and I promise not to force push again.)

Hopefully there aren't any remaining major issues. But issues may exist for sure.

cc: @lesteve @adrinjalali @thomasjpfan 

Thanks!! and Happy New Year!

adrinjalali commented on sklearn/utils/_sparse.py:
```diff
@@ -0,0 +1,174 @@
+
+########### fixes for transitioning function names
+
+# TODO: Replace when Scipy 1.12 is the minimum supported version
```
this should rather be in `fixes.py` to keep most version dependent code in one place

adrinjalali reviewed (commented):
I like this, but there's one thing worrying me, which is how are we going to make sure we don't use old matrices in new code that we write? Is there a way to get errors in one CI job if any old matrix is used? That would make it much easier for me to merge this PR.

Thanks for the work so far @dschult

dschult committed:
Merge branch 'main' into impl_as_sparse_function

dschult committed:
Update versionadded for sparse_interfac in _config.py

dschult committed:
Parametrize tests over allo_unlabeled and sparse_feature

dschult committed:
move sparse fixes functions to utils.fixes.py

dschult commented:
I made the suggested changes (moving the functions in fixes.py produced churn within the PR but mostly for imports.)

I understand your question about how to test that new code still works with the spmatrix inputs. After this PR, all "internal" code will use sparray. And new functions/code should do that same thing. Then it also needs to check input sparse objects for spmatrix and convert before further processing. The output should be controled by the configuration setting. So new code will need 3 things:
- check sparse input for spmatrix and convert if so.  Something like:
```python
    if issparse(X):
        if sp.sparse.isspmatrix(X):
            X = sp.sparse.csr_array(X)  # or slightly more complicated if format is unsure
        # other sparse input processing
```
- all further new code should work with this sparray object.
- set any returned sparse objects using the configuration setting `sklearn.utils._align_api_if_sparse()` (even if returned as part of a larger object).

> Is there a way to get errors in one CI job if any old matrix is used? 
I'm not sure what you are looking for here. Some `conftest.py` setting that would run all sparse tests using spmatrix input? Or maybe you mean the tool I used to find all the required changes: in conftest.py I monkeypatched scipy.sparse to make every spmatrix method call to e.g. `__mul__` raise. I only used that locally. Are you thinking of a CI run with this sort of check?

dschult committed:
Merge branch 'main' into impl_as_sparse_function

lorentzenchr commented:
@adrinjalali How about merging this and adding a sparse matrix sniffing test in a follow up PR. It can be a pain to maintain this PR for a longer time because it touches so many places.

dschult committed:
Add spatrix test for new code to sklearn ensuring no old idioms

isolating in its own commit for easy cherry-picking if needed

dschult committed:
Fixes to tests due to check_spmatrix option for pytest

skip tests with pickle and spmatrix.
skip tests that look for spmatrix.__module__
update `class is other_class` idiom to `issubclass` function

dschult committed:
add private function to handle case when pytest is not installed

dschult committed:
Merge branch 'main' into impl_as_sparse_function

dschult commented:
The requested pytest option to test for usage of spmatrix idioms has been committed.

Please let me know what else should be done here.
I am also looking for suggestions about where to put this spmatrix test within the CI machinery.
@adrinjalali requested "a way to get errors in one CI job if any old matrix is used". Where should that be?

Thanks!

dschult committed:
Merge branch 'main' into impl_as_sparse_function

lorentzenchr commented:
I would strongly recommend following the 80-20 rule with best effort PRs. We won‘t deliver 100% error free code. And that is fine for me as long as we are committed to invest in fixing those bugs afterwards. In fact, it is more efficient - in an *agile* reasoning - to deliver early and iterate.
(That‘s actually more than my 2 Cents.)

dschult committed:
detailed review of code involving sparse onjects

dschult committed:
Merge branch 'main' into impl_as_sparse_function

thomasjpfan commented on sklearn/conftest.py:
```diff
@@ -319,6 +424,17 @@ def pytest_generate_tests(metafunc):
     if not PARALLEL_RUN_AVAILABLE:
         parser.addini("thread_unsafe_fixtures", "list of stuff")
+    parser.addoption(
+        "--check_spmatrix",
```
Is the plan to run this in CI at some point?

thomasjpfan commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -177,6 +177,7 @@ def test_agglomerative_clustering_distances(
         assert not hasattr(clustering, "distances_")
 
 
+@pytest.mark.no_check_spmatrix  # pickle breaks check_spmatrix
```
How will a contributor know to add the `no_check_spmatrix` marker?

thomasjpfan reviewed (commented):
I'm +1 on moving forward with this before the next release. High level thoughts:

1. I suspect there will be unexpected side effects from this PR that we will uncover when it's released.
2. It'll be easy to miss the `_align_api_if_sparse` for future functions.

dschult committed:
Merge branch 'main' into impl_as_sparse_function

dschult commented:
Sounds great -- I merged with main. I'll watch for review questions --  :)

dschult commented:
> I will try to a detailed review next week, do ping me if you don't hear back from me!

Hi @lesteve! Just a quick ping to remind you about the detailed review of this PR. Happy March!

adrinjalali committed:
Merge branch 'main' into impl_as_sparse_function

lesteve reviewed (commented):
Here is the promised detailed review (better late than never 😅)! Below are the comments that may make sense to tackle before this PR is merged.

I may have a few other comments, but they can likely be done in follow-up PRs. The hope would be to merge this PR and iterate if follow-up PRs if needed.

dschult committed:
address review comments and suggestions
